---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [50]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [11]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [13]:
# G1 = P1_Graphs[0]

# for g in P1_Graphs:
#     print(nx.info(g))

Name: barabasi_albert_graph(1000,2)
Type: Graph
Number of nodes: 1000
Number of edges: 1996
Average degree:   3.9920
Name: watts_strogatz_graph(1000,10,0.05)
Type: Graph
Number of nodes: 1000
Number of edges: 5000
Average degree:  10.0000
Name: watts_strogatz_graph(750,5,0.075)
Type: Graph
Number of nodes: 750
Number of edges: 1500
Average degree:   4.0000
Name: barabasi_albert_graph(750,4)
Type: Graph
Number of nodes: 750
Number of edges: 2984
Average degree:   7.9573
Name: watts_strogatz_graph(750,4,1)
Type: Graph
Number of nodes: 750
Number of edges: 1500
Average degree:   4.0000


In [17]:
# for g in P1_Graphs:
#     ana(g)

1000 : 4.099161161161161 : 0.03167539146454044
1000 : 5.089871871871872 : 0.5642419635919628
750 : 9.378702269692925 : 0.4018222222222227
750 : 3.1048046283934134 : 0.03780379975223251
750 : 5.0785509568313305 : 0.0033037037037037037


In [15]:
# def ana(G1):
#     d1 = G1.degree()
    
#     deg = sorted(set(d1.values()))

#     #[list(d1.values()).count(i)/float(nx.number_of_nodes(G1)) for i in deg]
    
#     print(nx.number_of_nodes(G1),nx.average_shortest_path_length(G1), nx.average_clustering(G1) , sep = ' : ')

In [18]:
def graph_identification():
    
    # Your Code Here
    
    return ['PA','SW_L','SW_L','PA','SW_H']# Your Answer Here

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [4]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

# Steps:
## 1. Preprocessing
    >> Create the DF based on the nodes with employee_id as the index
    >> Extract useful features from the graph
    >> Separate the data based on management_salary column
## 2. EDA
## 3. Model
    Try Random Forest, GBTs, SVM, log_reg, KNNs
    If default ROC > 75 try grid search or manual tuning  
    Else move on to next model

In [3]:
# df = pd.DataFrame(index=G.nodes())
# df['Department'] = pd.Series(nx.get_node_attributes(G ,'Department'))
# df['Deg_cen'] = pd.Series(nx.degree_centrality(G).values())[0]
# df['Clo_cen'] = pd.Series(nx.closeness_centrality(G).values())[0]
# df['Bet_cen'] = pd.Series(nx.betweenness_centrality(G,k=100).values())[0]
# df['Target'] = pd.Series(nx.get_node_attributes(G ,'ManagementSalary'))
# df.head(5)

# df_train = df[~np.isnan(df['Target'])].copy(deep = True)
# df_test = df[np.isnan(df['Target'])].copy(deep = True)


# X_train = df_train.drop('Target',axis = 1)
# Y_train = df_train['Target']

# X_test = df_test.drop('Target',axis = 1)

# from sklearn.metrics import roc_auc_score
# from sklearn.model_selection import GridSearchCV
# from sklearn.ensemble import RandomForestClassifier

# model = RandomForestClassifier(random_state = 1,n_jobs = -1)

# params ={'n_estimators' : [6,8,10,12,14] , 'max_depth' : [2,3,4,5,6] }

# mod_sel = GridSearchCV(model,params,scoring = 'roc_auc')

# mod_sel.fit(X_train,y = Y_train)

# #mod_sel.best_params_

# mod = mod_sel.best_estimator_

# y_pred = mod.predict_proba(X_train)

# #roc_auc_score(Y_train , y_pred[:,1])

# ans = mod.predict_proba(X_test)[:,1]

# fin_df = pd.Series(data = ans,index=X_test.index)

# fin_df

Department   Deg_cen   Clo_cen   Bet_cen  Target
0           1  0.043825  0.421991  0.000664     0.0
1           1  0.051793  0.422360  0.000840     NaN
2          21  0.094622  0.461490  0.004923     NaN
3          21  0.070717  0.441663  0.001390     1.0
4          21  0.095618  0.462152  0.007857     1.0

In [9]:
def salary_predictions():
    
    # Your Code Here
    df = pd.DataFrame(index=G.nodes())
    df['Department'] = pd.Series(nx.get_node_attributes(G ,'Department'))
    df['Deg_cen'] = pd.Series(nx.degree_centrality(G).values())[0]
    df['Clo_cen'] = pd.Series(nx.closeness_centrality(G).values())[0]
    df['Bet_cen'] = pd.Series(nx.betweenness_centrality(G,k=100).values())[0]
    df['Target'] = pd.Series(nx.get_node_attributes(G ,'ManagementSalary'))
    df.head(5)

    df_train = df[~np.isnan(df['Target'])].copy(deep = True)
    df_test = df[np.isnan(df['Target'])].copy(deep = True)


    X_train = df_train.drop('Target',axis = 1)
    Y_train = df_train['Target']

    X_test = df_test.drop('Target',axis = 1)

    from sklearn.metrics import roc_auc_score
    from sklearn.model_selection import GridSearchCV
    from sklearn.ensemble import RandomForestClassifier

    model = RandomForestClassifier(random_state = 1,n_jobs = -1,n_estimators = 12,max_depth = 3)

    #params ={'n_estimators' : [6,8,10,12,14] , 'max_depth' : [2,3,4,5,6] }

    #mod_sel = GridSearchCV(model,params,scoring = 'roc_auc')

    model.fit(X_train,Y_train)

    #mod_sel.best_params_

    mod = model#.best_estimator_

    y_pred = mod.predict_proba(X_train)

    #roc_auc_score(Y_train , y_pred[:,1])

    ans = mod.predict_proba(X_test)[:,1]

    fin_df = pd.Series(data = ans,index=X_test.index)
    
    return fin_df# Your Answer Here

In [11]:
# a = salary_predictions()

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [74]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [76]:
# future_connections['common_neighbours'] = future_connections.index.map(lambda city: len(list(nx.common_neighbors(G, city[0], city[1]))))

# future_connections['preferential_attachment'] = [i[2] for i in nx.preferential_attachment(G, future_connections.index)]

# future_connections['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G, future_connections.index)]

# future_connections['resource_allocation_index'] = [i[2] for i in nx.resource_allocation_index(G, future_connections.index)]

# future_connections['adamic_adar_index'] = [i[2] for i in nx.adamic_adar_index(G, future_connections.index)]

# df = future_connections.copy(deep = True)

# cols = list(df)
# # move the column to head of list using index, pop and insert
# cols.insert(len(cols), cols.pop(cols.index('Future Connection')))
# # use ix to reorder
# df = df.ix[:, cols]

# test = df[np.isnan(df['Future Connection'])]
# train = df[~np.isnan(df['Future Connection'])]

# X_train = train.drop('Future Connection',axis=1)

# y_train = train['Future Connection']

# X_test = test.drop('Future Connection',axis=1)

# from sklearn.metrics import roc_auc_score
# from sklearn.model_selection import GridSearchCV
# from sklearn.ensemble import RandomForestClassifier

# model = RandomForestClassifier(random_state = 1,n_jobs = -1,max_depth= 9, n_estimators = 14)

# model.fit(X_train,y_train)

# y_pred = model.predict_proba(X_test)[:,1]

# ans = pd.Series(data = y_pred, index = X_test.index)

# ans

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:17: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated


(107, 348)    0.038146
(542, 751)    0.012592
(20, 426)     0.540290
(50, 989)     0.012433
(942, 986)    0.012894
(324, 857)    0.012664
(13, 710)     0.117300
(19, 271)     0.138830
(319, 878)    0.012217
(659, 707)    0.012592
(49, 843)     0.012420
(208, 893)    0.012592
(377, 469)    0.013857
(405, 999)    0.020420
(129, 740)    0.013919
(292, 618)    0.055522
(239, 689)    0.012066
(359, 373)    0.014373
(53, 523)     0.028766
(276, 984)    0.012420
(202, 997)    0.012630
(604, 619)    0.089432
(270, 911)    0.012217
(261, 481)    0.063164
(200, 450)    0.872375
(213, 634)    0.012592
(644, 735)    0.182533
(346, 553)    0.012237
(521, 738)    0.012022
(422, 953)    0.016935
                ...   
(672, 848)    0.012217
(28, 127)     0.989435
(202, 661)    0.012448
(54, 195)     1.000000
(295, 864)    0.012664
(814, 936)    0.012592
(839, 874)    0.012894
(139, 843)    0.012664
(461, 544)    0.013806
(68, 487)     0.013806
(622, 932)    0.012592
(504, 936)    0.018206
(479, 528) 

In [63]:
# from sklearn.metrics import roc_auc_score
# from sklearn.model_selection import GridSearchCV
# from sklearn.ensemble import RandomForestClassifier

# model = RandomForestClassifier(random_state = 1,n_jobs = -1)

# params ={'n_estimators' : [6,10,14,18,22,24,26] , 'max_depth' : [9,11,13,15] }
# mod_sel = GridSearchCV(model,params,scoring = 'roc_auc')

# mod_sel.fit(X_train,y = y_train)

# mod_sel.best_params_

#  1st n_ est [6,10,14,18,22,24,26] max_dep = [3,5,7,9]  = 14,9


In [9]:
def new_connections_predictions():
    
    # Your Code Here
    future_connections['common_neighbours'] = future_connections.index.map(lambda city: len(list(nx.common_neighbors(G, city[0], city[1]))))

    future_connections['preferential_attachment'] = [i[2] for i in nx.preferential_attachment(G, future_connections.index)]

    future_connections['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G, future_connections.index)]

    future_connections['resource_allocation_index'] = [i[2] for i in nx.resource_allocation_index(G, future_connections.index)]

    future_connections['adamic_adar_index'] = [i[2] for i in nx.adamic_adar_index(G, future_connections.index)]

    df = future_connections.copy(deep = True)

    cols = list(df)
    # move the column to head of list using index, pop and insert
    cols.insert(len(cols), cols.pop(cols.index('Future Connection')))
    # use ix to reorder
    df = df.ix[:, cols]

    test = df[np.isnan(df['Future Connection'])]
    train = df[~np.isnan(df['Future Connection'])]

    X_train = train.drop('Future Connection',axis=1)

    y_train = train['Future Connection']

    X_test = test.drop('Future Connection',axis=1)

    from sklearn.metrics import roc_auc_score
    from sklearn.model_selection import GridSearchCV
    from sklearn.ensemble import RandomForestClassifier

    model = RandomForestClassifier(random_state = 1,n_jobs = -1,max_depth= 9, n_estimators = 14)

    model.fit(X_train,y_train)

    y_pred = model.predict_proba(X_test)[:,1]

    ans = pd.Series(data = y_pred, index = X_test.index)

    return ans # Your Answer Here

In [30]:
# ind = future_connections.index